drivers/serial/usart_gd32.c

```diff
@@ -93,25 +148,201 @@ static int usart_gd32_err_check(const struct device *dev)
+	return 0;
+}
+
+/** Interrupt driven FIFO read function */
+int usart_gd32_fifo_read(const struct device *dev, uint8_t *rx_data,
+			const int size)
+{
+	const struct gd32_usart_config *const cfg = dev->config;
+	uint8_t num_rx = 0U;
+
+	while ((size - num_rx > 0) &&
+		usart_flag_get(cfg->reg, USART_FLAG_RBNE)) {
+		rx_data[num_rx++] = usart_data_receive(cfg->reg);
+	}
+
+	return 0;
+}
+
+/** Interrupt driven transfer enabling function */
+void usart_gd32_irq_tx_enable(const struct device *dev)
+{
+	const struct gd32_usart_config *const cfg = dev->config;
+
+	usart_interrupt_enable(cfg->reg, USART_INT_TC);
+}
+
+/** Interrupt driven transfer disabling function */
+void usart_gd32_irq_tx_disable(const struct device *dev)
+{
+	const struct gd32_usart_config *const cfg = dev->config;
+
+	usart_interrupt_disable(cfg->reg, USART_INT_TC);
+}
+
+/** Interrupt driven transfer ready function */
+int usart_gd32_irq_tx_ready(const struct device *dev)
+{
+	const struct gd32_usart_config *const cfg = dev->config;
+
+	return usart_flag_get(cfg->reg, USART_FLAG_TBE) &&
+		usart_interrupt_flag_get(cfg->reg, USART_INT_FLAG_TC);
+}
+
+/** Interrupt driven transfer complete function */
+int usart_gd32_irq_tx_complete(const struct device *dev)
+{
+	const struct gd32_usart_config *const cfg = dev->config;
+
+	return usart_flag_get(cfg->reg, USART_FLAG_TC);
+}
+
+/** Interrupt driven receiver enabling function */
+void usart_gd32_irq_rx_enable(const struct device *dev)
+{
+	const struct gd32_usart_config *const cfg = dev->config;
+
+	usart_interrupt_enable(cfg->reg, USART_INT_RBNE);
+}
+
+/** Interrupt driven receiver disabling function */
+void usart_gd32_irq_rx_disable(const struct device *dev)
+{
+	const struct gd32_usart_config *const cfg = dev->config;
+
+	usart_interrupt_disable(cfg->reg, USART_INT_RBNE);
+}
+
+/** Interrupt driven receiver ready function */
+int usart_gd32_irq_rx_ready(const struct device *dev)
+{
+	const struct gd32_usart_config *const cfg = dev->config;
+
+	return usart_flag_get(cfg->reg, USART_FLAG_RBNE);
+}
+
+/** Interrupt driven error enabling function */
+void usart_gd32_irq_err_enable(const struct device *dev)
+{
+	const struct gd32_usart_config *const cfg = dev->config;
+
+	usart_interrupt_enable(cfg->reg, USART_INT_ERR);
+	usart_interrupt_enable(cfg->reg, USART_INT_PERR);
+}
+
+/** Interrupt driven error disabling function */
+void usart_gd32_irq_err_disable(const struct device *dev)
+{
+	const struct gd32_usart_config *const cfg = dev->config;
+
+	usart_interrupt_disable(cfg->reg, USART_INT_ERR);
+	usart_interrupt_disable(cfg->reg, USART_INT_PERR);
+}
+
+/** Interrupt driven pending status function */
+int usart_gd32_irq_is_pending(const struct device *dev)
+{
+	const struct gd32_usart_config *const cfg = dev->config;
+
+	return (usart_flag_get(cfg->reg, USART_FLAG_RBNE) &&
+		usart_interrupt_flag_get(cfg->reg, USART_INT_RBNE)) ||
+		(usart_flag_get(cfg->reg, USART_FLAG_TC) &&
+		usart_interrupt_flag_get(cfg->reg, USART_INT_TC));
```

Comment:
wrong flags, it won't works:

```
	return ((usart_flag_get(cfg->reg, USART_FLAG_RBNE) &&
		 usart_interrupt_flag_get(cfg->reg, USART_INT_FLAG_RBNE)) ||
		(usart_flag_get(cfg->reg, USART_FLAG_TC) &&
		 usart_interrupt_flag_get(cfg->reg, USART_INT_FLAG_TC)));
```

Reply:
Thanks! It's easy to mix the INT_FLAG and FLAG.